src/Voron/Impl/Compaction/StorageCompaction.cs

```diff
@@ -230,57 +230,97 @@ private static void CopyTrees(StorageEnvironment existingEnv, StorageEnvironment
+                            foreach (var result in inputTable.SeekForwardFrom(index, lastSlice))
+                            {
+                                foreach (var entry in result.Results)
+                                {
+                                    // The table will take care of reconstructing indexes automatically
+                                    outputTable.Insert(entry);
+                                    entriesLeft--;
+                                    transactionSize += entry.Size;
+                                }
+
+                                lastSlice = result.Key;
+
+                                // The transaction has surpassed the allowed
+                                // size before a flush
+                                if (transactionSize >= compactedEnv.Options.MaxScratchBufferSize / 2)
+                                    break;
+                            }
+                        }
+                        else
                         {
-                            foreach (var entry in result.Results)
+                            // Use a fixed size index
+                            var index = schema.FixedSizeIndexes.First().Value;
+
+                            foreach (var entry in inputTable.SeekForwardFrom(index, lastFixedIndex))
                             {
+
                                 // The table will take care of reconstructing indexes automatically
                                 outputTable.Insert(entry);
+                                entriesLeft--;
+                                transactionSize += entry.Size;
+                                lastFixedIndex = index.GetValue(entry);
+
+                                // The transaction has surpassed the allowed
+                                // size before a flush
+                                if (transactionSize >= compactedEnv.Options.MaxScratchBufferSize / 2)
+                                    break;
                             }
                         }
                     }
                     else
                     {
-                        // Use a fixed size index
-                        var index = schema.FixedSizeIndexes.First().Value;
-
-                        foreach (var entry in inputTable.SeekForwardFrom(index, long.MinValue))
+                        // The table has a primary key, inserts in that order are expected to be faster
+                        foreach (var entry in inputTable.SeekByPrimaryKey(lastSlice))
                         {
-
                             // The table will take care of reconstructing indexes automatically
                             outputTable.Insert(entry);
+                            entriesLeft--;
+                            transactionSize += entry.Size;
+                            lastSlice = schema.Key.GetSlice(txr.Allocator, entry);
```

Comment:
Here you are copying the slice _each and every time_, instead of only when we need to break because of size issues.